query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+        match engine
+            .executor()
+            .start_tx(engine.query_schema().clone(), engine.engine_protocol(), tx_opts)
+            .instrument(span)
+            .await
+        {
+            Ok(tx_id) => Ok(json!({ "id": tx_id.to_string() }).to_string()),
+            Err(err) => Ok(map_known_error(err)?),
+        }
+        // }
+        // .with_subscriber(dispatcher)
+        // .await
+    }
+
+    // If connected, attempts to commit a transaction with id `tx_id` in the core.
+    pub async fn commit_transaction(&self, tx_id_str: *const c_char, _trace: *const c_char) -> Result<String, Error> {
+        let tx_id = get_cstr_safe(tx_id_str).expect("Input string missing");
+        let inner = self.inner.read().await;
+        let engine = inner.as_engine()?;
+
+        // let dispatcher = self.logger.dispatcher();
+
+        // async move {
+        match engine.executor().commit_tx(TxId::from(tx_id)).await {
+            Ok(_) => Ok("{}".to_string()),
+            Err(err) => Ok(map_known_error(err)?),
+        }
+        // }
+        // .with_subscriber(dispatcher)
+        // .await
+    }
+
+    // If connected, attempts to roll back a transaction with id `tx_id` in the core.
+    pub async fn rollback_transaction(&self, tx_id_str: *const c_char, _trace: *const c_char) -> Result<String, Error> {
+        let tx_id = get_cstr_safe(tx_id_str).expect("Input string missing");
+        // let trace = get_cstr_safe(trace_str).expect("trace is required in transactions");
+        let inner = self.inner.read().await;
+        let engine = inner.as_engine()?;
+
+        // let dispatcher = self.logger.dispatcher();
+
+        // async move {
+        match engine.executor().rollback_tx(TxId::from(tx_id)).await {
+            Ok(_) => Ok("{}".to_string()),
+            Err(err) => Ok(map_known_error(err)?),
+        }
+        // }
+        // .with_subscriber(dispatcher)
+        // .await
+    }
+}
+
+//            _____ _____
+//      /\   |  __ \_   _|
+//     /  \  | |__) || |
+//    / /\ \ |  ___/ | |
+//   / ____ \| |    _| |_
+//  /_/    \_\_|   |_____|
+//
+// This API is meant to be stateless. This means the box pointer to the query engine structure will be returned to the
+// calling code and should be passed to subsequent calls
+//
+// Be should be careful about not de-allocating the pointer
+// when adding a new function remember to always call mem::forget
+
+/// # Safety
+///
+/// The calling context needs to pass a valid pointer that will store the reference
+#[no_mangle]
+pub unsafe extern "C" fn prisma_create(options: ConstructorOptions, qe_ptr: *mut *mut QueryEngine) -> c_int {
+    if qe_ptr.is_null() {
+        return PRISMA_MISSING_POINTER;
+    }
+
+    let res = QueryEngine::new(options);
+    match res {
+        Ok(v) => {
+            *qe_ptr = Box::into_raw(Box::new(v));
+            PRISMA_OK
+        }
+        Err(e) => {
+            println!("Error creating query engine {:#?}", e);
```

Comment:
oh yeah, I forgot about this one, thanks!

Reply:
Ah, actually here the trick doesn't work, because I have nowhere to store the underlaying string (the struct has not been created). But again, I'm open on how to handle this more gracefully. Maybe creating a string and putting it in a Box is the correct way to do this? I just don't know how to deal with memory leaks due to Boxes being left everywhere.

Reply:
Solved by https://github.com/prisma/prisma-engines/pull/4774/files#r1535395731